### DISCUSSION TITLE
Long-term simulation (multiple years) using a predifined current profile

### DISCUSSION THREAD

**Discussion by jhthompson12** [pybamm-team/PyBaMM#4152] [Q&A] [answered]
Hi! 

This is a follow up to a previous question: "[Modeling the operation of a cell from real world energy storage system data](https://github.com/pybamm-team/PyBaMM/discussions/4147)". Essentially, I have 6 years of current data from a real lithium ion NMC cell (LG Chem JH3) that I would like to run through some PyBaMM model / simulation to see how the model's discharge capacity degrades over time. 

Since receiving help in that thread [I have had success getting  a model to run on short periods of data (hours to days)](https://github.com/pybamm-team/PyBaMM/discussions/4147#discussioncomment-9705459) but as i try to extend my simulation I run into a predictable issue where the simulated cell's voltage creeps up over time until it hits the upper voltage limit which stops the simulation: 

![image](https://github.com/pybamm-team/PyBaMM/assets/45108842/86566b82-a492-4e3f-ab63-152663639908)

This growing error between modeled and real seems to mainly come from some self discharge / loss of voltage that the real cell experiences over time that the model does not express.

In the [simulating-long-experiments notebook](https://github.com/pybamm-team/PyBaMM/blob/develop/docs/source/examples/notebooks/simulations_and_experiments/simulating-long-experiments.ipynb) I found the `starting_solution` attribute from `simulation.solve(starting_solution=sol)` which seems like it could be useful here. But in that documentation it is used in the context of repeating a regular experiment many times over.

One way to address my issue could be to break my years of cycling data into more manageable simulations, for example daily, where I feed the last day's solution into the next. But critically I would need to reset the `initial_soc` to realign the model to the real voltage in each new simulation. 

So, my questions are:
- is this possible? I guess I should just try providing both  `starting_solution` and `initial_soc` to a simulation and see what happens? 
- each day's simulation would have to update the current data that it is simulating. is it possible to pass on most of `starting_solution`s information while overwriting the current profile that should be used in the next simulation? 

Thank you in advance!

**Comment by TomTranter** [comment]
Instead of updating the current function parameter you could use an experiment step for each day. https://docs.pybamm.org/en/stable/source/api/experiment/experiment_steps.html

**Reply by TomTranter** [reply]
I would also make a new simulation for each day and write results to files to clear down memory

**Reply by jhthompson12** [reply]
@TomTranter 

I took a stab at incorporating your suggestions above and am getting this error on the 2nd day's simulation (i.e. the first simulation where I try using both `starting_solution` and `initial_soc`):

```
[ERROR] callbacks.on_experiment_error(224): Simulation error: Solver failed because the following interpolation bounds were exceeded: ["Interpolant 'interpolating_function' lower bound"]. You may need to provide additional interpolation points outside these bounds.
```

Can you please help me to understand what the error is and what I can do to address it? 

### Here is 3 days of sample data where my error is occurring when trying to simulate the 2nd day:
[sample_data_3_days.csv](https://github.com/user-attachments/files/15777160/sample_data_3_days.csv)

### Here is my code:

``` python 
"""
This file attempts to address the issue identified in this discussion: https://github.com/pybamm-team/PyBaMM/discussions/4152#discussioncomment-9707173

Let's try to take the very long timeseries data that we would like to simulate and break it up into a smaller chain of
sequential simulations each feeding into the next. Hopefully using [experiment_steps](https://docs.pybamm.org/en/stable/source/api/experiment/experiment_steps.html)?

When feeding one simulation into the next try to realign the modeled vs real cell voltages by using `initial_soc`
.solve() argument.
"""

import pybamm
import matplotlib as mpl
import pandas as pd  # needed to read the csv data file

pybamm.set_logging_level("VERBOSE")
mpl.use('TkAgg')  # or can use 'TkAgg', whatever you have/prefer

LG_JH3_NOM_AH = 63

# extract smaller portion of data
start = pd.to_datetime("2018-12-14 00:00:00")
end = pd.to_datetime("2018-12-17 00:00:00")

# The rack from which this data originates is made of 17 modules (in series), where each module contains 28 cells in a
# 14S2P configuration. So, to get the cell-level current we divide the rack-level current by 2
# create a C-Rate profile so that we can create the corresponding current profile for a different capacity cell
# create a seconds column
smaller_data = pd.read_csv("sample_data_3_days.csv")

# Instantiate a simple model
model = pybamm.lithium_ion.DFN()

# get predefined parameters for an NMC cell
param_mohtat = pybamm.ParameterValues("Mohtat2020")

# SoC / Voltage relationship data from real system
param_mohtat["Open-circuit voltage at 0% SOC [V]"] = 3.2
param_mohtat["Open-circuit voltage at 100% SOC [V]"] = 4.17
param_mohtat["Upper voltage cut-off [V]"] = 4.2
# nominal capacity for JH3 cells (and a few other parameters) from LG Chem documentation
# param_mohtat["Nominal cell capacity [A.h]"] = 63      # changing the nominal cell capacity will have little effect. You need to adjust electrode dimensions to change actual capacity.
# param_mohtat["Cell volume [m3]"] = 5.65e-04
param_mohtat["Lower voltage cut-off [V]"] = 3

output_variables = [
    'Terminal voltage [V]',
    'Voltage [V]',
    'Current [A]',
    'Ambient temperature [K]',
    "Discharge capacity [A.h]",
    "Total lithium capacity [A.h]"
]

# loop over each day to create the chain of simulations
all_days_sim_data = pd.DataFrame()
for day in pd.date_range(start=start, end=end, freq="d")[:-1]:
    print(day)
    day_start, day_end = day, day + pd.Timedelta(days=1)
    day_data = smaller_data.loc[day_start:day_end].iloc[:-1]    # drop off last timestamp measure which will be first of next day

    # create interpolant - must be a function of *dimensional* time
    drive_cycle = day_data[["Seconds", "Cell Current"]]
    drive_cycle.loc[:, "Cell Current"] = day_data["C-Rate"] * param_mohtat["Nominal cell capacity [A.h]"]  # convert C-Rate to current for modeled cell
    drive_cycle = drive_cycle.to_numpy()
    day_experiment = pybamm.Experiment([pybamm.step.current(drive_cycle)])

    # set up simulation - for drive cycles we recommend using the CasadiSolver in "fast" mode
    # solver = pybamm.CasadiSolver(mode="fast")   # was running into errors with this one, so leaving out for now
    day_simulation = pybamm.Simulation(model, experiment=day_experiment, parameter_values=param_mohtat)

    initial_soc = day_data['Container 1, Rack 1_StateOfCharge'].iloc[0] / 100
    try:                        # previous day's solution exists, feed it into the next simulation
        day_sol = day_simulation.solve(starting_solution=day_sol, initial_soc=initial_soc - (initial_soc * 0.1), showprogress=True)     # Apply a -10% correction factor for SoC to get the voltages to align better
    except NameError as e:      # day_sol does not exist because this is the first loop
        day_sol = day_simulation.solve(initial_soc=initial_soc - (initial_soc * 0.1), showprogress=True)                                # Apply a -10% correction factor for SoC to get the voltages to align better

    # extract this day's solution data and append to the larger multi-day data for plotting
    this_day_sim_data = pd.concat(
        [pd.Series(day_sol[var].entries, index=day_sol.t, name=var) for var in output_variables],
        axis=1
    )

    all_days_sim_data = pd.concat([all_days_sim_data, this_day_sim_data])
```

Thanks in advance for any advice!

**Reply by jhthompson12** [reply]
I think i found the issue, it had to do with how i was specifying the time values in experiments for the days after the first. 

Initially I was picking up where the time left off from the previous experiment, so day 1's experiment had time values from 0 to 86400 and day 2 would have 86400 to 172800. But starting the experiment at a not zero seems to have been the culprit. 

When I changed the daily experiments to always have time values specified from 0 to 86400 everything ran fine without the `interpolation bounds were exceeded` error. And the solver seemed to naturally pick up the solution at the timestamp where it left off.

**Reply by jhthompson12** [reply]
@TomTranter 

> I would also make a new simulation for each day and write results to files to clear down memory

Im not sure I am properly executing what you describe here.. 

This is my code now ([the same data from my comment above](https://github.com/user-attachments/files/15777160/sample_data_3_days.csv) should still be applicable): 

``` python
"""
This file attempts to address the issue identified in this discussion: https://github.com/pybamm-team/PyBaMM/discussions/4152#discussioncomment-9707173

Let's try to take the very long timeseries data that we would like to simulate and break it up into a smaller chain of
sequential simulations each feeding into the next. Hopefully using [experiment_steps](https://docs.pybamm.org/en/stable/source/api/experiment/experiment_steps.html)?

When feeding one simulation into the next try to realign the modeled vs real cell voltages by using `initial_soc`
.solve() argument.
"""

import pybamm
import matplotlib as mpl
import pandas as pd  # needed to read the csv data file

pybamm.set_logging_level("VERBOSE")
mpl.use('TkAgg')  # or can use 'TkAgg', whatever you have/prefer

LG_JH3_NOM_AH = 63

# extract smaller portion of data
start = pd.to_datetime("2018-12-14 00:00:00")
end = pd.to_datetime("2018-12-17 00:00:00")

# The rack from which this data originates is made of 17 modules (in series), where each module contains 28 cells in a
# 14S2P configuration. So, to get the cell-level current we divide the rack-level current by 2
# create a C-Rate profile so that we can create the corresponding current profile for a different capacity cell
# create a seconds column
smaller_data = pd.read_csv("sample_data_3_days.csv")

# Instantiate a simple model
model = pybamm.lithium_ion.DFN()

# get predefined parameters for an NMC cell
param_mohtat = pybamm.ParameterValues("Mohtat2020")

# SoC / Voltage relationship data from real system
param_mohtat["Open-circuit voltage at 0% SOC [V]"] = 3.2
param_mohtat["Open-circuit voltage at 100% SOC [V]"] = 4.17
param_mohtat["Upper voltage cut-off [V]"] = 4.2
# nominal capacity for JH3 cells (and a few other parameters) from LG Chem documentation
# param_mohtat["Nominal cell capacity [A.h]"] = 63      # changing the nominal cell capacity will have little effect. You need to adjust electrode dimensions to change actual capacity.
# param_mohtat["Cell volume [m3]"] = 5.65e-04
param_mohtat["Lower voltage cut-off [V]"] = 3

output_variables = [
    'Terminal voltage [V]',
    'Voltage [V]',
    'Current [A]',
    'Ambient temperature [K]',
    "Discharge capacity [A.h]",
    "Total lithium capacity [A.h]"
]

# loop over each day to create the chain of simulations
all_days_sim_data = pd.DataFrame()
for day in pd.date_range(start=start, end=end, freq="d")[:-1]:
    print(day)
    day_start, day_end = day, day + pd.Timedelta(days=1)
    day_data = smaller_data.loc[day_start:day_end]#.iloc[:-1]    # drop off last timestamp measure which will be first of next day

    # create a seconds column for just this day
    day_data.loc[:, "Seconds"] = (day_data.index - day_data.index[0]).total_seconds()

    # create interpolant - must be a function of *dimensional* time
    drive_cycle = day_data[["Seconds", "Cell Current"]]
    drive_cycle.loc[:, "Cell Current"] = day_data["C-Rate"] * param_mohtat["Nominal cell capacity [A.h]"]  # convert C-Rate to current for modeled cell
    drive_cycle = drive_cycle.to_numpy()
    day_experiment = pybamm.Experiment([pybamm.step.current(drive_cycle)])

    # set up simulation - for drive cycles we recommend using the CasadiSolver in "fast" mode
    solver = pybamm.CasadiSolver(mode="fast")   # was running into errors with this one, so leaving out for now
    day_simulation = pybamm.Simulation(model, experiment=day_experiment, parameter_values=param_mohtat)#, solver=solver)

    initial_soc = day_data['Container 1, Rack 1_StateOfCharge'].iloc[0] / 100
    try:                        # previous day's solution exists, feed it into the next simulation
        day_sol = day_simulation.solve(starting_solution=day_sol, initial_soc=initial_soc - (initial_soc * 0.1), showprogress=True)     # Apply a -10% correction factor for SoC to get the voltages to align better
    except NameError as e:      # day_sol does not exist because this is the first loop
        day_sol = day_simulation.solve(initial_soc=initial_soc - (initial_soc * 0.1), showprogress=True)                                # Apply a -10% correction factor for SoC to get the voltages to align better

    # extract this day's solution data and append to the larger multi-day data for plotting
    this_day_sim_data = pd.concat(
        [pd.Series(day_sol[var].entries, name=var) for var in output_variables] + [pd.Series(day_sol.t, name="Seconds")],
        axis=1
    )
    
    all_days_sim_data = pd.concat([all_days_sim_data, this_day_sim_data])
```

I feel like i am "making a new simulation for each day", and from the solver's DEBUG output it seems like each day's solver is truly only running calculations for the next 86400 seconds in each iteration, but something does not seem quite right.. Maybe it's the fact that each time I access the data for a day's simulation object I am seeing ALL data from the past days, which does make sense i am picking up where the last solution left off, but I feel like I am maybe not getting the memory clearing benefits you mentioned with this approach? Should I wait until all day's simulations are complete before accessing the simulation data for plotting? 

Are you able to verify that I am on the right track with what you were suggesting?

**Reply by jhthompson12** [reply]
Im also not sure if `initial_soc` is working the way I was hoping it would here either. The daily simulation approach does seem to make it a bit further than the full 10-day simulation (it is able to complete the simulation for the 19th then fails on the 20th whereas it would previously stop simulating the 18th)

<img width="1412" alt="Screenshot 2024-06-10 at 3 58 40 PM" src="https://github.com/pybamm-team/PyBaMM/assets/45108842/f0f272ee-cb5d-45d0-b8ee-c541edfa32b9">

I was hoping that providing `initial_soc` at the beginning of each day would sort of realign the SoC of the modeled cell with the real cell buying me more time before the upper voltage limit was reached. This would result in a step change in the cell voltage at the beginning of each day, maybe this is not possible for the model?

**Reply by TomTranter** [reply]
So I think you want to first focus on getting good fits for your RPT cycles then look at fitting over time. You can adjust the capacity nominally by changing electrode height and width but looking at your pulses it seems like you haven't got a terribly good fit for the rest of the parameters which would be important to get better values for. You would want to get a good idea of the amount of active material and cyclable lithium and also the OCVs and exchange currents, particle sizes and diffusion coefficients for the cell at BOL at a bare minimum for an SPM model. If you are struggling fitting BOL there's very little chance this will work just feeding in current traces. A simpler first approach to model degradation would be to fit the model for each rpt and manually adjust the active material and lithium inventory to see what the actually cap fade is and see whether the rest of the characteristics like pulse resistance are also lining up, then you have an idea of which degradation mechanism might be dominating.

**Reply by jhthompson12** [reply]
Thanks, Tom! 

Unfortunately, I think that fitting the model to my real system data kind of defeats the purpose of my exploration. Also, I not sure I could find all of that proprietary information about this particular cell without doing a tear down or testing in a lab setting. 

One hack I just tried: instead of feeding my real system current profile to the simulations, use the real system's rack-level average cell voltage with `pybamm.Experiment([pybamm.step.voltage(drive_cycle)])`

This allowed me to make it through the 10 day simulation without getting tripped up by the model's voltage limits being reached. The resulting current profile from the simulation ends up looking pretty much exactly like the C-Rate that my real cell experiences. 

![image](https://github.com/pybamm-team/PyBaMM/assets/45108842/012c75e4-a361-4988-94d1-cdba463588b0)

This approach will hopefully be enough to allow me to simulate the many years of operation of my real system in PyBaMM to at least extract a degradation profile from the model.. but I'll have to see how this scales past just 10 days. At the moment each day's simulation is taking ~2 minutes. At that rate it would take around 3 days to run through the simulations for my data!

**Reply by lappemic** [reply]
@TomTranter is the from @jhthompson12 provided codesnippet the suggested way to simulate longer experiments? As stated in [this discussion](https://github.com/pybamm-team/PyBaMM/discussions/4212) i fail to simulate longer experiments and it would really help me get this working.

**Reply by jhthompson12** [reply, marked as answer]
@lappemic, what I posted above did end up working for me, but I made some tweaks to the approach to make things work more smoothly. This is very similar is spirit, but changed slightly to take advantage of PyBaMM's `experiment_step` workflow. 

With this I was able to run a 7 year simulation at 10 second resolution broken into daily experiment steps (i.e. ~2500 "cycles") which took around 3 hours to run.  

``` python
"""
This file attempts to address the issue identified in this discussion: https://github.com/pybamm-team/PyBaMM/discussions/4152#discussioncomment-9707173

Let's try to take the very long timeseries data that we would like to simulate and break it up into a smaller chain of
sequential simulations each feeding into the next. Hopefully using [experiment_steps](https://docs.pybamm.org/en/stable/source/api/experiment/experiment_steps.html)?
"""

import pybamm
import matplotlib as mpl
import pandas as pd  # needed to read the csv data file

pybamm.set_logging_level("NOTICE")
mpl.use('TkAgg')  # or can use 'TkAgg', whatever you have/prefer

LG_JH3_NOM_AH = 63

start = pd.to_datetime("2018-12-14 00:00:00")
end = pd.to_datetime("2018-12-17 00:00:00")

# import drive cycle data from file
all_real_rack_data = pd.read_pickle("sample_data_3_days.csv")

# Instantiate a simple model
model = pybamm.lithium_ion.SPM(options={"SEI": "reaction limited"})

# get predefined parameters for an NMC cell
param_mohtat = pybamm.ParameterValues("Mohtat2020")

# SoC / Voltage relationship data from real system
param_mohtat["Open-circuit voltage at 0% SOC [V]"] = 3.2
param_mohtat["Open-circuit voltage at 100% SOC [V]"] = 4.17
param_mohtat["Upper voltage cut-off [V]"] = 4.2
# nominal capacity for JH3 cells (and a few other parameters) from LG Chem documentation
# param_mohtat["Nominal cell capacity [A.h]"] = 63      # changing the nominal cell capacity will have little effect. You need to adjust electrode dimensions to change actual capacity.
# param_mohtat["Cell volume [m3]"] = 5.65e-04
param_mohtat["Lower voltage cut-off [V]"] = 3

output_variables = [
    'Voltage [V]',
    'Current [A]',
]

# create a separate experiment step for each day to create the chain of daily simulations
experiment_steps = []
all_days_sim_data = pd.DataFrame()
for day in pd.date_range(start=start, end=end, freq="d")[:-1]:
    # print("Simulating: ", day)
    day_start, day_end = day, day + pd.Timedelta(days=1)
    day_data = smaller_data.loc[day_start:day_end].copy()#.iloc[:-1]    # drop off last timestamp measure which will be first of next day

    # create a seconds column for just this day
    day_data.loc[:, "Seconds"] = (day_data.index - day_data.index[0]).total_seconds()

    # create this day's experiment step
    drive_cycle = day_data[["Seconds", "Container 1, Rack 1_DC.Voltage_Avg of Cells"]]
    drive_cycle = drive_cycle.to_numpy()
    experiment_steps.append(pybamm.step.voltage(drive_cycle))

experiment = pybamm.Experiment(experiment_steps)

# set up simulation - for drive cycles we recommend using the CasadiSolver in "fast" mode
solver = pybamm.CasadiSolver(mode="fast")   # was running into errors with this one, so leaving out for now
day_simulation = pybamm.Simulation(model, experiment=experiment, parameter_values=param_mohtat)#, solver=solver)

initial_soc = day_data['Container 1, Rack 1_StateOfCharge'].iloc[0] / 100
cycles_to_save = list(range(0, len(experiment_steps) + 1, 60))[1:] + [len(experiment_steps)]    # Save a cycle each 60 days including the first and last
select_sol = day_simulation.solve(initial_soc=initial_soc, showprogress=True, save_at_cycles=cycles_to_save)

pybamm.plot_summary_variables(select_sol)
```

Apologies if this script does not immediately run for you. I may have broken a couple things while trying to clean it up for posting, but the ideas within worked for me.

**Reply by lappemic** [reply]
Perfect, thanks a lot @jhthompson12! This really helped! I adapted your logic and simulate now for batches of 100s. This seems to be better memorywise, it still increases and i am testing up to how many cycles i get. Thanks again for posting here!
